include/deal.II/differentiation/sd/symengine_scalar_operations.h

```diff
@@ -130,9 +135,676 @@ namespace Differentiation
 
     //@}
 
+    /**
+     * @name Symbolic map creation
+     */
+    //@{
+
+    namespace internal
+    {
+      /**
+       * Return whether or not an @p entry is a valid symbol that
+       * we can expect to perform substitution on.
+       */
+      bool
+      is_valid_substitution_symbol(const SE::Basic &entry);
+    } // namespace internal
+
+    //@}
+
+    /**
+     * @name Symbolic substitution map enlargement
+     */
+    //@{
+
+    namespace internal
+    {
+      /**
+       * A convenience function for an adding an entry to the substitution map
+       * @p substitution_map. The new entry will have the key given by @p symbol
+       * with its paired @p value. Such maps are required to perform substitution
+       * of symbolic expressions, with key entries being exchanges with their
+       * pair values.
+       *
+       * There are cases where its convenient to simply ignore the fact that we
+       * may be trying to add invalid symbols. For example, one may wish to add
+       * a tensor where only some entries are symbols and the others are zero'd.
+       * In this case we ensure that the user knows what they're doing by
+       * forcing them to override this safety mechanism with a template
+       * parameter. These types of functions are typically called in a manner
+       * indicating that the user knows exactly what they're passing into it.
+       *
+       * @tparam ignore_invalid_symbols A template parameter that enforces whether
+       *         or not the @p symbol has to be a valid one or not. In the
+       *         overwhelming majority of cases, the default value of
+       *         <tt>false</tt> should be selected, with the result that an
+       *         exception will be thrown if the input @p symbolic is, in fact,
+       *         not a symbolic value or expression.
+       */
+      template <bool ignore_invalid_symbols = false>
+      void
+      add_to_substitution_map(types::substitution_map &       substitution_map,
+                              const SE::RCP<const SE::Basic> &symbol,
+                              const SE::RCP<const SE::Basic> &value);
+    } // namespace internal
+
+    /**
+     * A convenience function for an adding an entry to the substitution map
+     * @p substitution_map. The new entry will have the key given by @p symbol
+     * with its paired @p value. Such maps are required to perform substitution
+     * of symbolic expressions, with key entries being exchanges with their
+     * pair values.
+     *
+     * This function is guarenteed to create an ordering that is identical
+     * to the typical add_to_symbol_map() call that may used initially
```

Comment:
This should have read

> This function is guaranteed to create an ordering that is identical to the typical add_to_symbol_map() call that may be used when initially configuring a BatchOptimizer.

I've fixed it now. Thanks for being permissive in my mentioning both the `add_to_symbol_map()` function and `BatchOptimizer` thingy. These, of course, will both come into existence in future PRs.

Reply:
Let me elucidate on the role of these two things so that you can decide if this sentence is indeed clear. 

The `BatchOptimizer` will take a whole collection of symbolic expressions, preprocess them, and (in some manner) optimally evaluate them all at once. So when one first constructs a `BatchOptimizer` one needs to know only symbolic variables and expressions, and nothing numeric. That's what the `add_to_symbol_map()` call can do: its make a map like a `type::substitution_map`, but with a dummy `value` for each `key` (the dummy being hidden from the user). Later, one uses `add_to_substitution_map()` to create the full substitution map from scratch, or a function `set_value_in_symbol_map()` to set the value entries created using `add_to_symbol_map()`. So these all manipulate a substitution map, but help give some context as to what you want to do with the map and when you want to perform these operations on it.